### ISSUE TITLE
The out-of-bounds array access behavior is inconsistent with that of the VCS.

### ISSUE THREAD

**Issue by Superchao-ai** [verilator/verilator#2984]
Hi,
I‘m using the release version of verilator-4.202 and here is my example:
`
module top(clk);
	input clk;

	logic [15:0] index;
	logic [15:0] data;
	logic [15:0] mem[7:0];

	always @(posedge clk)
		begin
			data  <= data + 1'b1;
			index <= index + 1'b1;
			mem[index] <= data;  // index will be out of range
		end

endmodule
`

I notice that the verilator output cpp is as follows:
`
VL_INLINE_OPT void Vtop::_sequent__TOP__3(Vtop__Syms* __restrict vlSymsp) {
    VL_DEBUG_IF(VL_DBG_MSGF("+    Vtop::_sequent__TOP__3\n"); );
    Vtop* const __restrict vlTOPp VL_ATTR_UNUSED = vlSymsp->TOPp;
    // Variables
    // Begin mtask footprint all: 
    CData/*2:0*/ __Vdlyvdim0__top__DOT__mem__v0;
    SData/*15:0*/ __Vdly__top__DOT__data;
    SData/*15:0*/ __Vdly__top__DOT__index;
    SData/*15:0*/ __Vdlyvval__top__DOT__mem__v0;
    // Body
    __Vdly__top__DOT__index = vlTOPp->top__DOT__index;
    __Vdly__top__DOT__data = vlTOPp->top__DOT__data;
    __Vdly__top__DOT__data = (0xffffU & ((IData)(1U) 
                                         + (IData)(vlTOPp->top__DOT__data)));
    __Vdly__top__DOT__index = (0xffffU & ((IData)(1U) 
                                          + (IData)(vlTOPp->top__DOT__index)));
    __Vdlyvval__top__DOT__mem__v0 = vlTOPp->top__DOT__data;
    __Vdlyvdim0__top__DOT__mem__v0 = (7U & (IData)(vlTOPp->top__DOT__index));
    vlTOPp->top__DOT__data = __Vdly__top__DOT__data;
    vlTOPp->top__DOT__index = __Vdly__top__DOT__index;
    vlTOPp->top__DOT__mem[__Vdlyvdim0__top__DOT__mem__v0] 
        = __Vdlyvval__top__DOT__mem__v0;
}
`
`__Vdlyvdim0__top__DOT__mem__v0 = (7U & (IData)(vlTOPp->top__DOT__index));`
this code shows that only the lower 3-bit used as the index.
If the index exceeds the depth of the array, will incorrectly assigned to the array.
In VCS simulation, indexes that exceed the array depth are ignored and no value is assigned.

Can you solve this problem by keeping the simulation behavior of the verilator consistent with that of the VCS?

**Comment by wsnyder** [comment]
The behavior you want is required by IEEE, so this is a bug.

Verilator does simplify the range checks for performance, I need to look into this case to see if this case can be properly handled without significant performance loss.  I suspect not.

**Comment by PeterMonsson** [comment]
I remember this as a simulation vs. synthesis mismatch issue. Consider adding a warning for this case.

**Comment by gezalore** [comment]
FWIW, compilers seem to turn `a >= C  ? C : a` into a branch free sequence with `cmov` (on x86), or `csel` (on arm):
https://godbolt.org/z/TW8GfcKvx
so it might not be too bad if you just make all arrays 1 larger than declared, and dump out of range values in the spare entry:
```
vlTOPp->top__DOT__mem[__Vdlyvdim0__top__DOT__mem__v0 >= 8 ? 8 : __Vdlyvdim0__top__DOT__mem__v0] = ...
```

**Comment by margej** [comment]
I suspect many synthesis tools behave in the same way here. 
Rather than forfeit performance for strict IEEE compliance, which may not match a synthesised design, as @PeterMonsson said, I think a lint warning may be a better approach?

**Comment by wsnyder** [comment]
I suspect a new warning, but no time has been put into this recently.  If you or someone would like to look into it, that would be appreciated.

**Comment by jmichalski-ant** [comment]
It looks like this problem was at least partially addressed, because now there is a warning when the design posted above is used:
```
%Warning-WIDTHTRUNC: oob.v:62:7: Bit extraction of array[7:0] requires 3 bit index, not 16 bits.
                               : ... note: In instance 'top'
   62 |    mem[index] <= data;   
      |       ^
                     ... For warning description see https://verilator.org/warn/WIDTHTRUNC?v=5.049
                     ... Use "/* verilator lint_off WIDTHTRUNC */" and lint_on around source to disable this message.
```
but in the cases where the size of the indexed array is not a power of 2, there is no warning and there is an out-of-bounds access in the generated c++ code.  
For example the following code results in a segfault:
```
module t();
  reg clk = 0;
  wire w;
  reg [15:0] idx = 65535;
  reg m[0:34000];
  assign w = m[idx];

  always @(posedge clk)
  begin
    idx <= idx - 1;
    m[idx] <= 0;
  end

  initial begin
    clk = ~clk;
    $display("%d", w);
  end
endmodule
```
```
$ ./bin/verilator oob.v --binary
$ ./obj_dir/Voob
Segmentation fault
```
This is also reported by sanitizer, just note that `-fstrict-flex-arrays=1` has to be used, otherwise it assumes that the array in `VlUnpacked` is a flexible one and skips bound checks on it:
```
$ ./bin/verilator oob.v --binary -CFLAGS "-fstrict-flex-arrays=1 -fsanitize=undefined" -LDFLAGS "-fsanitize=undefined"
$ ./obj_dir/Voob
/home/ant/verilator/include/verilated_types.h:1468:48: runtime error: index 65535 out of bounds for type 'Unpacked' (aka 'unsigned char[34001]')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/ant/verilator/include/verilated_types.h:1468:48 
```
While making reliable warning in such case might be hard to implement (either we would have to track all index variable changes and determine if any `m[idx]` access will be with `idx` value out-of-bounds or it would give false-positives), maybe we can
at least modify the wrapping-around behavior to use modulo-wrap instead of masking lower bits when the size of the array is not a power of two, to avoid out-of-bounds accesses in c++ code?

**Comment by gezalore** [comment]
Agreed this could be improved. In general we can't warn if the index is not known at compile time.

> we can at least modify the wrapping-around behavior to use modulo-wrap instead of masking lower bits when the size of the array is not a power of two, to avoid out-of-bounds accesses in c++ code?

Masking is used for speed, a modulo would need a division on every access which is slow. IEEE says a write to an OOB location is a no-op. Verilator tries to do that but there are bugs. If you are interested in making a pull, V3Unknown is the pass that tries to insert the bounds checking.

**Comment by jmichalski-ant** [comment]
`V3Unknown` was already inserting bound checks in this case, but they were later misoptimized. I made a PR with fix for that: https://github.com/verilator/verilator/pull/7755
